public/app/features/plugins/components/AppRootPage.tsx

```diff
@@ -81,16 +82,56 @@ export function AppRootPage({ pluginId, pluginNavSection }: Props) {
     />
   );
 
+  // Because of the fallback at plugin routes, we need to check
+  // if the user has permissions to see the plugin page.
+  const userHasPermissionsToPluginPage = (path: string) => {
+    if (!plugin.meta?.includes) {
+      return true;
+    }
+    let rolesMap = new Map();
+    plugin.meta?.includes.map((include) => {
+      rolesMap.set(include.path, include.role);
+    });
+    if (contextSrv.isGrafanaAdmin) {
+      return true;
+    }
+    if (contextSrv.isEditor && rolesMap.get(path) === 'Viewer') {
+      return true;
+    }
+    return contextSrv.hasRole(rolesMap.get(path));
+  };
+
+  const AccessDenied = () => {
+    return (
+      <Alert severity="warning" title="Access denied">
+        You do not have permission to see this plugin.
+      </Alert>
+    );
+  };
   if (!pluginNav) {
-    return <PluginPageContext.Provider value={context}>{pluginRoot}</PluginPageContext.Provider>;
+    return (
+      <>
+        {userHasPermissionsToPluginPage(pluginRoot.props.path) ? (
+          <PluginPageContext.Provider value={context}>{pluginRoot}</PluginPageContext.Provider>
+        ) : (
+          <AccessDenied />
+        )}
+      </>
+    );
   }
 
   return (
     <>
       {navModel ? (
-        <Page navModel={navModel} pageNav={pluginNav?.node}>
-          <Page.Contents isLoading={loading}>{pluginRoot}</Page.Contents>
-        </Page>
+        <>
+          {userHasPermissionsToPluginPage(pluginRoot.props.path) ? (
+            <Page navModel={navModel} pageNav={pluginNav?.node}>
+              <Page.Contents isLoading={loading}>{pluginRoot}</Page.Contents>
+            </Page>
+          ) : (
+            <AccessDenied />
+          )}
+        </>
```

Comment:
I am afraid that this will only block access in case there is no `navModel`, otherwise it will render `<Page>{pluginRoot}</Page>` even if the user has no access. Am I correct?